From 44185f0ccb00402bfd79c4082991e71f794ac3c5 Mon Sep 17 00:00:00 2001
From: Sergio Pena <sergio.pena@cloudera.com>
Date: Thu, 17 Nov 2016 12:43:26 -0600
Subject: [PATCH 0820/1164] HIVE-15226: Add a different masking comment to
 qtests blobstore output (Sergio Pena, reviewed by
 Mohit Sabharwal and Sahil Takiar)

Change-Id: I8c0a3e43387f4a3dfeee95701c763c38e9be8c40
---
 itests/hive-blobstore/pom.xml                      |   11 +++++++++
 .../test/results/clientpositive/insert_into.q.out  |    4 ++--
 .../control/AbstractCoreBlobstoreCliDriver.java    |    2 ++
 .../java/org/apache/hadoop/hive/ql/QTestUtil.java  |   24 ++++++++++++++++++--
 4 files changed, 37 insertions(+), 4 deletions(-)

diff --git a/itests/hive-blobstore/pom.xml b/itests/hive-blobstore/pom.xml
index aaccdc9..b119502 100644
--- a/itests/hive-blobstore/pom.xml
+++ b/itests/hive-blobstore/pom.xml
@@ -62,6 +62,17 @@
         <maven.test.skip>false</maven.test.skip>
       </properties>
     </profile>
+    <profile>
+      <id>hadoop-2</id>
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hbase</groupId>
+          <artifactId>hbase-server</artifactId>
+          <version>${hbase.hadoop2.version}</version>
+          <scope>test</scope>
+        </dependency>
+      </dependencies>
+    </profile>
   </profiles>
 
   <dependencies>
diff --git a/itests/hive-blobstore/src/test/results/clientpositive/insert_into.q.out b/itests/hive-blobstore/src/test/results/clientpositive/insert_into.q.out
index 919f3e7..93f545d 100644
--- a/itests/hive-blobstore/src/test/results/clientpositive/insert_into.q.out
+++ b/itests/hive-blobstore/src/test/results/clientpositive/insert_into.q.out
@@ -4,12 +4,12 @@ POSTHOOK: query: DROP TABLE qtest
 POSTHOOK: type: DROPTABLE
 #### A masked pattern was here ####
 PREHOOK: type: CREATETABLE
-PREHOOK: Input: #### A masked pattern was here ####
+PREHOOK: Input: ### test.blobstore.path ###/qtest
 PREHOOK: Output: database:default
 PREHOOK: Output: default@qtest
 #### A masked pattern was here ####
 POSTHOOK: type: CREATETABLE
-POSTHOOK: Input: #### A masked pattern was here ####
+POSTHOOK: Input: ### test.blobstore.path ###/qtest
 POSTHOOK: Output: database:default
 POSTHOOK: Output: default@qtest
 PREHOOK: query: INSERT INTO qtest VALUES (1), (10), (100), (1000)
diff --git a/itests/util/src/main/java/org/apache/hadoop/hive/cli/control/AbstractCoreBlobstoreCliDriver.java b/itests/util/src/main/java/org/apache/hadoop/hive/cli/control/AbstractCoreBlobstoreCliDriver.java
index 1e7fd5e..e12e3e4 100644
--- a/itests/util/src/main/java/org/apache/hadoop/hive/cli/control/AbstractCoreBlobstoreCliDriver.java
+++ b/itests/util/src/main/java/org/apache/hadoop/hive/cli/control/AbstractCoreBlobstoreCliDriver.java
@@ -162,5 +162,7 @@ private void setupUniqueTestPath() {
     String uid = new SimpleDateFormat("yyyyMMdd.HHmmss.SSS").format(Calendar.getInstance().getTime())
         + "-" + String.format("%03d", (int)(Math.random() * 999));
     testBlobstorePathUnique = testBlobstorePath + uid;
+
+    qt.addPatternWithMaskComment(testBlobstorePathUnique, String.format("### %s ###", HCONF_TEST_BLOBSTORE_PATH));
   }
 }
diff --git a/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java b/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java
index 69d3144..45cde4a 100644
--- a/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java
+++ b/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestUtil.java
@@ -71,6 +71,8 @@
 import org.apache.commons.io.IOUtils;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
@@ -1481,6 +1483,18 @@ private void maskPatterns(Pattern[] patterns, String fname) throws Exception {
       }
 
       if (!partialMaskWasMatched) {
+        for (Pair<Pattern, String> pair : patternsWithMaskComments) {
+          Pattern pattern = pair.getLeft();
+          String maskComment = pair.getRight();
+
+          matcher = pattern.matcher(line);
+          if (matcher.find()) {
+            line = matcher.replaceAll(maskComment);
+            partialMaskWasMatched = true;
+            break;
+          }
+        }
+
         for (Pattern pattern : patterns) {
           line = pattern.matcher(line).replaceAll(maskPattern);
         }
@@ -1539,8 +1553,7 @@ private void maskPatterns(Pattern[] patterns, String fname) throws Exception {
       ".*Output:.*/data/files/.*",
       ".*total number of created files now is.*",
       ".*.hive-staging.*",
-      "org\\.apache\\.hadoop\\.hive\\.metastore\\.model\\.MConstraint@([0-9]|[a-z])*",
-      "(s3.?|swift|wasb.?):\\/\\/[\\w\\.\\/-]*"
+      "org\\.apache\\.hadoop\\.hive\\.metastore\\.model\\.MConstraint@([0-9]|[a-z])*"
   });
 
   private final Pattern[] partialReservedPlanMask = toPattern(new String[] {
@@ -1548,6 +1561,13 @@ private void maskPatterns(Pattern[] patterns, String fname) throws Exception {
       //TODO: add more expected test result here
   });
 
+  /* This list may be modified by specific cli drivers to mask strings that change on every test */
+  private final List<Pair<Pattern, String>> patternsWithMaskComments = new ArrayList<>();
+
+  public void addPatternWithMaskComment(String patternStr, String maskComment) {
+    patternsWithMaskComments.add(ImmutablePair.of(Pattern.compile(patternStr), maskComment));
+  }
+
   public int checkCliDriverResults(String tname) throws Exception {
     assert(qMap.containsKey(tname));
 
-- 
1.7.9.5

